### PR TITLE
brcm63xx: add Comtrend VR-3022eu support

### PR DESCRIPTION
**Pull request by mruzicka**
Add support for Comtrend VR-3022eu.
The device is very similar to the Comtrend VR-3025u, except it only has an 8MiB flash. While featuring a smaller flash, its PCB is more populated adding a 1Gbps port to the integrated switch via an external PHY, a second USB port, WLAN and WPS GPIO buttons and a WPS LED.

To support 1Gbps external PHY reliably a patch for the bcm63xx_enet driver was necessary, please see the commit adding the patch for details.

Specification:
- SoC: Broadcom BCM6368
- CPU: BMIPS4350 / 400 MHz (dual core)
- RAM: 64 MiB
- Flash: 8 MiB
- Wired Ethernet:
  - 1x 1Gbps Port
  - 4x 100Mbps Port
  - All ports connected to the integrated BCM6368 Roboswitch (b53) switch
- WLAN: Broadcom BCM43222, 2x non-detachable internal antenna
- USB: 2x USB 2.0 Port
- UART: plated through hole pads on the PCB, header needs to be soldered
- Buttons: Reset, WLAN Switch, WPS
- LEDs:
  - GPIO Controlled: Power (Reg/Green), Inet (Reg/Green), WPS (Green), DSL (Green)
  - HW Controlled: Internal 100Mbps PHY 1-4 (Green), External 1Gbps PHY (Red/Green), WLAN (Green)

Flash Instruction:
1. Connect your PC to one of the four 100Mbps ethernet ports of the device. Don't use the 1Gbps port!
2. Power up the device with the reset button pressed and keep it pressed for about 10 seconds.
3. Setup a static IP on your PC: 192.168.1.2/24
4. Visit the 192.168.1.1 webpage in your browser and follow the instruction there to upload / flash a new firmware.

Signed-off-by: Michal Růžička <michal.ruza@gmail.com>

### CODE REVIEW

File: target/linux/brcm63xx/base-files/etc/board.d/02_network [hunk 1]
```diff
@@ -98,6 +98,11 @@ vr-3026e)
 		"0:lan:1" "1:lan:2" "2:lan:3" "3:lan:4" "8t@eth0"
 	;;
 
+vr-3022eu)
+	ucidef_add_switch "switch0" \
+		"0:lan:1" "1:lan:2" "2:lan:3" "3:lan:4" "5:lan:17" "8t@eth0"
```
Comment (mruzicka) [inline target/linux/brcm63xx/base-files/etc/board.d/02_network]: I'm not quite sure what the port index is good for. Apparently here I set it equal to the phy_id of the PHY connected to the respective Roboswitch port which is probably completely bogus.

File: target/linux/brcm63xx/patches-4.14/581-board_VR-3022eu.patch [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+ };
+ 
++static struct sprom_fixup __initdata vr3022eu_fixups[] = {
++	{ .offset = 97, .value = 0xfeb3 },
++	{ .offset = 98, .value = 0x1618 },
++	{ .offset = 99, .value = 0xfab0 },
++	{ .offset = 113, .value = 0xfed1 },
++	{ .offset = 114, .value = 0x1609 },
++	{ .offset = 115, .value = 0xfad9 },
++};
```
Comment (mruzicka) [inline target/linux/brcm63xx/patches-4.14/581-board_VR-3022eu.patch]: I copied these from the `vr3025u_fixups`. I don't know how to asses whether it's correct or not. The BCM43222 seems to work though.

**Comment by mruzicka** [comment]
I've yet updated the 805-bcm63xx_enet_disabled_port_override_support.patch patches to print the "invalid forced speed on port" warning if port's bypass_link == 1 && force_speed == 0 which was the original behavior.

File: target/linux/brcm63xx/patches-4.14/581-board_VR-3022eu.patch (hunk 2 above)
Comment (KanjiMonster) [reply target/linux/brcm63xx/patches-4.14/581-board_VR-3022eu.patch]: You should find these in the appropriate GPL tarball. There should be a 43222_map.bin or so for the values.

File: target/linux/brcm63xx/base-files/etc/board.d/02_network (hunk 1 above)
Comment (KanjiMonster) [reply target/linux/brcm63xx/base-files/etc/board.d/02_network]: You labeled the port as "wan", so it should be "5:wan" here as well.

**Comment by KanjiMonster** [comment]
> The 1Gb port would not transmit any data if a 1Gb link was negotiated while the respective port speed override in the integrated Roboswitch was set to anything else than 1Gb. To fix that the bcm63xx_enet driver is updated to apply the port's force_speed value (when specified in the board parameters) to the port when it is being disabled. Note that this can only happen when bypass_link is not specified for the port at the same time. This is then taken advantage of by setting the force_speed of the affected port to 1000 in the board parameters which makes the port functional at any negotiated link speed.

This all sounds a bit strange to me.

1. This patch should be a separate commit.
2. You shouldn't need to force any speed. Does the port even work at lower speeds if you force it to 1000?
3. You are doing several things in this patch, e.g. resetting the ports as well or doing renames of functions. Please split this up and drop the rename.

File: target/linux/brcm63xx/base-files/etc/board.d/02_network (hunk 1 above)
Comment (mruzicka) [reply target/linux/brcm63xx/base-files/etc/board.d/02_network]: I labeled it as "wan" in the board params as that's how it's labeled on the router itself. IIUC labeling it "wan" here assigns it the wan _role_ which may or may not be desired.

My thinking was that given it's the only 1Gb port of the router one may want to use it for the presumably faster internal network while using one of the FastEthernet ports as the "wan" port. If this doesn't sound right I'm happy to change it.

Comment (KanjiMonster) [reply target/linux/brcm63xx/base-files/etc/board.d/02_network]: I try to go by the principle of least surprise; if the stock firmware assigns this port as a wan port (and is labeled as such), then OpenWrt should do the same by default.

File: target/linux/brcm63xx/patches-4.14/581-board_VR-3022eu.patch (hunk 2 above)
Comment (mruzicka) [reply target/linux/brcm63xx/patches-4.14/581-board_VR-3022eu.patch]: There are the following files in the GPL tarball (which may not the tarball for the exact router model, but probably close):
/bcmdrivers/broadcom/net/wl/impl8/bcm4322_map.bin_save
/bcmdrivers/broadcom/net/wl/impl8/bcm43222_map.bin_save
/bcmdrivers/broadcom/net/wl/impl8/bcm43224_map.bin_save
/bcmdrivers/broadcom/net/wl/impl8/bcm43225_map.bin_save
/bcmdrivers/broadcom/net/wl/impl8/bcm43226_map.bin_save

In addition to that, there are these files in the original firmware's filesystem:
/etc/wlan/bcm4322_map.bin
/etc/wlan/bcm43222_map.bin
/etc/wlan/bcm43224_map.bin
/etc/wlan/bcm43225_map.bin
/etc/wlan/bcm43226_map.bin
/etc/wlan/bcm43227_map.bin
/etc/wlan/bcm43228_map.bin

The corresponding files from these two sources are the same except for bcm43225_map.bin_save/bcm43225_map.bin which differ in 12 bytes.

Which file do I need and how do I use it to verify the fixups are correct.

**Comment by mruzicka** [comment]
The 1G port behaves strangely only when you connect a 1Gb capable device to it. It works flawlessly (without the need to patch the bcm63xx_enet driver) if you only ever connect 10/100Mb devices.

But when you connect a 1Gb device then it seems it really depends on the value of Roboswitch's `ENETSW_PORTOV_REG(x)` register for the corresponding port at the time when the 1Gb links is established:
- If the `ENETSW_PORTOV_1000_MASK` bit is set in the register, then the port works at the 1Gb.
- If the bit is not set, then the port never transmits any data, not even the Roboswitch MIB counters of the port are updated. Strangely the port _can_ and _does_ receive traffic while in this weird state.

The `805-bcm63xx_enet_disabled_port_override_support.patch` solves that by setting the `ENETSW_PORTOV_1000_MASK` bit in the register when the port is being disabled which happens when the link goes down on the port (and also during initialization). This ensures that next time the link goes up, the bit is set and as that seems to be the necessary precondition (as described above) for the port to work even at the 1Gb, the port indeed does work even at that speed then. Note that the normal speed override still takes place when the link goes up on the port - for example when you connect a 100Mb device, the `ENETSW_PORTOV_1000_MASK` bit is cleared and the `ENETSW_PORTOV_100_MASK` is set in the register until the links goes down again, at which point the patched code makes sure the `ENETSW_PORTOV_1000_MASK` bit is set again so that the port is an a state which ensures it'll work even when you connect a 1Gb device next time.

I've spent quite some time trying to figure out what could be difference between the working and non working states of the port but couldn't really find anything in the Roboswitch registers and MII registers of the external PHY. I've added code which compared the values of all the Roboswitch control registers and relevant status registers and the MII registers of the external PHY for the two different states of the port but nothing really showed up. So I come to the conclusion the difference is not exposed through the Roboswitch and/or the PHY MII registers. But perhaps I have overlooked something, any thoughts?

File: target/linux/brcm63xx/base-files/etc/board.d/02_network (hunk 1 above)
Comment (mruzicka) [reply target/linux/brcm63xx/base-files/etc/board.d/02_network]: Fair enough.

File: target/linux/brcm63xx/patches-4.14/581-board_VR-3022eu.patch (hunk 2 above)
Comment (mruzicka) [reply target/linux/brcm63xx/patches-4.14/581-board_VR-3022eu.patch]: I found the templates in `arch/mips/bcm63xx/sprom.c`. It turns out the `/etc/wlan/bcm43222_map.bin` is quite similar to the `bcm43222_sprom` template, except the very end of the file:

* /etc/wlan/bcm43222_map.bin:
  ```
  ...
  0001a0 ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff
  0001b0 ff ff ff ff ff ff 44 08 00 00 00 00 00 00 00 00
  ```
* bcm43222_sprom:
  ```
  ...
  0001a0 ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff
  0001b0 ff ff ff ff ff ff 00 08
  ```
The `/etc/wlan/bcm43222_map.bin` is 8 bytes longer! (All those extra bytes are zeroes though.)
And there is a difference in the last but one byte of the `bcm43222_sprom` (0x44 vs. 0x00).

I'm wondering could that be some kind of a checksum or does that rather mean I need the following fixup:
```c
static struct sprom_fixup __initdata vr3022eu_fixups[] = {
        { .offset = 219, .value = 0x4408 },
};
```

Comment (mruzicka) [reply target/linux/brcm63xx/patches-4.14/581-board_VR-3022eu.patch]: I noticed the `EVG2000_fixups` from `arch/mips/bcm63xx/boards/board_bcm963xx.c` is quite similar to what I'm seeing. So hopefully the fixup I arrived at above makes sense.

**Comment by mruzicka** [comment]
@KanjiMonster I've updated the PR, I believe I've addressed all your comments. Please let me know if I missed anything.

File: target/linux/brcm63xx/dts/vr-3022eu.dts [hunk 3]
```diff
@@ -0,0 +1,108 @@
+/dts-v1/;
```
Comment (rmilecki) [inline target/linux/brcm63xx/dts/vr-3022eu.dts]: Please include licensing info for your DTS file, see
https://openwrt.org/submitting-patches#dts_checklist
e.g. a line like:
`// SPDX-License-Identifier: GPL-2.0-or-later OR MIT`
above the `/dts-v1/;`

Comment (ynezz) [reply target/linux/brcm63xx/dts/vr-3022eu.dts]: @mruzicka Are you still interested to get this in?

**Comment by adrianschmutzler** [comment]
Since there has not been any response from the author for quite some time, I close this PR.
Feel free to request reopening the PR when you continue to work on this. :-)

**Pull request closed without merging**